ENGLISH|[简体中文](README_CN.md)

<details><summary>Contents</summary>

<!-- TOC -->

- [Restraints Assisted Structure Predictor](#rasp)
- [iterative Folding Assisted peak ASsignmenT](#faast)
- [Environment](#environment)
  - [Hardware & Framework](#hardware--framework)
  - [Install RequireMents](#install--requirements)
- [Code Contents](#code-contents)
- [Example](#example)
  - [RASP Example](#rasp-inference)
  - [FAAST Example](#faast-method)
- [RunTime](#runtime)
- [Reference](#reference)
- [Acknowledgement](#acknowledgement)

<!-- /TOC -->

</details>

# RASP

Existing AI methods such as MEGA-Fold/AlphaFold have greatly improved the accuracy of static protein structure prediction. However, there are still unsolved problems, such as generating dynamic conformations and predicting the structure in accordance with experimental or prior information. In order to solve these problems, we developed the RASP (Restraints Assisted Structure Predictor) model. The architecture of RASP is derived from AlphaFold evoformer and structure module, and it accepts abstract or experimental restraints, sparse or dense, to generate structures. This enables RASP to be used for a variety of experiments and applications, and can improve the structure prediction of multi-domain proteins and proteins with few MSAs.

# FAAST

Nuclear magnetic resonance (NMR) is the only method that can resolve the conformation and dynamic structure of the solution state closer to the actual environment of proteins with atomic resolution [1][2]. However, the acquisition and analysis of NMR experimental data takes a long time. On average, a single protein requires experts to invest at least several months, most of which is spent on the analysis and attribution of experimental data. Existing NMR NOE peak data analysis methods such as CARA, ARIA, CYANA, etc., use the structure iterative analysis data generated by traditional molecular dynamics simulation. The analysis speed is slow, and the constraint information and structure analyzed from the data still need a lot of expert knowledge, and require a long time to further correction. In order to improve the speed and accuracy of NMR experimental data analysis, we developed FAAST (iterative Folding Assisted peak ASsignmenT), an automatic NMR data analysis method based on MindSpore+ Ascend AI software and hardware platform.

## Environment

### hardware--framework

RASP and FAAST are based on the biological computing library [MindSPONGE](<https://gitee.com/mindspore/mindscience/tree/master/MindSPONGE>) and MindSpore framework. They require MindSpore version 2.0 and later, which can be installed referring to [MindSpore installation page](https://www.mindspore.cn/install). This toolkit can be run on either Ascend910 or GPUs with memory no less than 16GB. The Ascend runtime calls mixed precision by default, and the GPU runtime uses full precision calculations.

### install--requirements

- Install MindSpore:
    Download mindspore wheel package:

    PlatForm | Link
    ----------|----------
    Ascend-910 ARM  | <https://ms-release.obs.cn-north-4.myhuaweicloud.com/2.0.0rc1/MindSpore/unified/aarch64/mindspore-2.0.0rc1-cp37-cp37m-linux_aarch64.whl>
    Ascend-910  x86 | <https://ms-release.obs.cn-north-4.myhuaweicloud.com/2.0.0rc1/MindSpore/unified/x86_64/mindspore-2.0.0rc1-cp37-cp37m-linux_x86_64.whl>
    GPU x86   | <https://ms-release.obs.cn-north-4.myhuaweicloud.com/2.0.0rc1/MindSpore/unified/x86_64/mindspore-2.0.0rc1-cp37-cp37m-linux_x86_64.whl>

    Installation on Ascend requires hardware driver package:Ascend Data Center Solution 23.0.RC1,[Install Guide](<https://www.mindspore.cn/install>)
    Installation on GPU requires Nvidia CUDA version 11.1-11.6. Please refer to the [Installation Link](<https://developer.nvidia.com/cuda-toolkit-archive>)

- Install MindSPONGE:
    Download Mindscience ：

    ``` shell
    git clone https://gitee.com/mindspore/mindscience.git
    ```

    compile MindSPONGE wheel package：

    ``` shell
    cd ./mindscience/MindSPONGE/
    ```

    if on Ascend910 platform

    ``` shell
    bash build.sh -e ascend -j 8
    ```

    if on GPU platform

    ``` shell
    bash build.sh -e gpu -j 8
    ```

    Install wheel package

    ``` shell
    pip install ./output/mindsponge*.whl
    ```

- Install other requirements：
    This tool relies on searching tools such as HHSearch and Kalign. A one-click installation script is provided for easy use.

    ``` shell
    cd ./mindscience/MindSPONGE/applications/research/FAAST
    sh ./install.sh
    ```

## code-contents

<details><summary><font size=4 color="blue">Code Contents</font></summary>

```bash
├── FAAST
    ├── main.py                            // FAAST Main Script
    ├── run_rasp.py                        // RASP Main Script
    ├── README.md                          // FAAST README English Version
    ├── README_CN.md                       // FAAST README Chinese Version
    ├── extract_restraints.py              // Extract The Constraint Sample File From The pdb
    ├── search.py                          // Call Mmseqs Online Search For MSA And Templates
    ├── install.sh                         // Install The Shell Script For Install Dependencies
    ├── assign
        ├── assign.py                      //Iterative Assignment
        ├── init_assign.py                 //Initial Assignment
    ├── commons
        ├── analysis.py                    //Result Analysis Tool
        ├── nmr_hydrogen_equivariance.txt  //Hydrogen Equivalence
        ├── res_constants.py               //Hydrogen Equivalence Dictionary
    ├── config
        ├── data.yaml                      //Data Process Config
        ├── model.yaml                     //Model Config
    ├── data
        ├── dataset.py                     // Dataset
        ├── hhsearch.py                    // HHsearch Tool
        ├── kalign.py                      // Kalign Tool
        ├── msa_query.py                   // MSA Processing Tool
        ├── parsers.py                     // Mmcif Processing
        ├── preprocess.py                  // Data Processing
        ├── protein_feature.py             // MSA and Template Feature Search and Integration script
        ├── templates.py                   // Template Processing Scripts
        ├── utils.py                       // Common Func Scripts
    ├── model
        ├── fold.py                        // Main RASP Model
    ├── module
        ├── evoformer.py                   // Evoformer Module
        ├── fold_wrapcell.py               // Wrapper
        ├── head.py                        // Heads
        ├── structure.py                   // Structure Module
        ├── template_embedding.py          // Template Module
    ├── nmr_relax
        ├── model  
            ├── structure_violation.py     //Calculate Whether The Structure Conforms To Physical Laws
            ├── utils.py                   //Common Tools For Relaxtion
        ├── relax
            ├── amber_minimize.py          //OpenMM Amber Force Field Minimization
            ├── cleanup.py                 //Cleanup
            ├── relax.py                   //Relaxation Scripts
            ├── utils.py                   //Common tools for Openmm
```

</details>

## example

### RASP inference

The RASP model trained weights can be downloaded at [RASP.ckpt](<https://download.mindspore.cn/mindscience/mindsponge/FAAST/checkpoint/RASP.ckpt>) and
the sample files can be downloaded in [example](<https://download.mindspore.cn/mindscience/mindsponge/FAAST/example/>)，running following command to start inference.

```bash
Usage:python run_rasp.py --run_platform PLATFORM --use_pkl False --restraints_path RESTRAINTS_PATH
            --input_path INPUT_FILE_PATH --checkpoint_file CHECKPOINT_FILE --use_template True --use_custom False
            --a3m_path A3M_PATH --template_path TEMPLATE_PATH

option:
--restraints_PATH    Path for the constraint information folder.Constraint information for each sample is stored in separate txt files with the same name as sequence/MSA.
--run_platform       The operating platform, can be Ascend or GPU
--input_path         The path for input .fasta/.pkl files. Multiple samples with different file names can be stored in one single folder.
--checkpoint_file    Model weight file path
--use_pkl            whether to use pkl file as input,default:False.
--use_template       whether to use template,default:True.
--use_custom         Whether to use custom MSA and template information provided by user, default:False.
--a3m_path           The MSA folder path for MSAs saved after searching, or custom MSAs provided by user.
--template_path      The template folder path for templates saved after searching, or custom templates provided by user.
```

The model supports three types of input. The first type of input is the protein fasta sequence, in which mode the MSA and template are generated by online mmseqs search. Enter a3m_path and template_path to save the search results.
The second type is to input the MSA and template files provided by the user, wherein MSA is in a3m format and template is in cif format. The MSA and templates can be searched with other tools by user or provided according to prior knowledge (e.g. known structure templates). To use this mode set use_pkl to False and use_custom to True, and enter the user-provided MSA and template paths a3m_path and template_path.
The third method is to use the pre-processed pkl file. use_pkl needs to be set to True. a3m_path and template path are not required.Pkl file preprocessing can refer to

```log
./data/protein_feature.py:monomer_feature_generate
```

It mainly deals with the characteristic information of the input sequence, the searched msa information and the template information, and saves the corresponding pkl file in./pkl_file/ after each run of RASP model.

**Restraint information**
The model requires constraint information as input, and the constraint information takes the form of a (N,2) text file describing which residue pairs are close in 3-dimentional space like `[[1,2],... [2,10]]`, etc. Constraint information comes from various sources, including mass spectrum cross-linking, fluorescent resonance energy transfer and so on. A sample script is provided to extract constraint information from the pdb, as shown below.

```bash
Usage:python extract_restraints.py --pdb_path PDB_PATH --output_file OUTPUT_FILE
option:
--pdb_path         A pdb file that provides constraint information.
--output_file      Location of the file to output constraint information.
```

The following is an example file of constraint information. Each line is the spatial location of a pair of amino acids. Each location is separated by a space.

``` log
51 74
46 60
.. ..
36 44
70 46
18 68
```

Inference results are stored in './result/ ', and each sequence generates the corresponding prot_name.pdb

```log
{confidence of predicted structrue :89.23, time :95.86，restraint recall :1.0}
```

img.png
For multi-domain structure 6XMV, both AlphaFold and MEGA-fold provide inaccurate relative domain positions, however with constraint RASP is able to fix the inter-domain structure.

### FAAST-METHOD

The RASP model trained weights can be downloaded at [RASP.ckpt](<https://download.mindspore.cn/mindscience/mindsponge/FAAST/checkpoint/RASP.ckpt>) and
the sample files can be downloaded in [example](<https://download.mindspore.cn/mindscience/mindsponge/FAAST/example/>)，running following command to start inference.

```bash
Usage: python main.py --run_platform PLATFORM --use_pkl True --peak_and_cs_path PEAKLIST_PATH
            --input_path INPUT_FILE_PATH --checkpoint_file CHECKPOINT_FILE --use_template True --use_custom False
            --a3m_path A3M_PATH --template_path TEMPLATE_PATH

option:
--peak_and_cs_path   Path of NOESY spectral data and chemical shift tables.
--run_platform       The operating platform, can be Ascend or GPU
--input_path         The path for input .fasta/.pkl files. Multiple samples with different file names can be stored in one single folder.
--checkpoint_file    Model weight file path
--use_pkl            whether to use pkl file as input,default:False.
--use_template       whether to use template,default:True.
--use_custom         Whether to use custom MSA and template information provided by user, default:False.
--a3m_path           The MSA folder path for MSAs saved after searching, or custom MSAs provided by user.
--template_path      The template folder path for templates saved after searching, or custom templates provided by user.
```

The input form supported by this method is the same as that of RASP model. Compared with RASP, it does not require constraint information, but requires chemical shift table and NOESY spectral peak data.

**NOESY peak list** : File name begins with 'noelist_', data must be converted to numpy.ndarray format and stored in .pkl file format, ndarray shape is' [num_peak, 4] ', as shown in the following example:

``` log
array([[3.257400e+01, 8.500000e+00, 4.878000e+00, 8.945950e+05],
       [4.557600e+01, 9.230000e-01, 3.922000e+00, 7.646940e+05],
       [3.148300e+01, 8.377000e+00, 5.340000e+00, 5.575200e+05],
       ...,
       [3.297700e+01, 3.166000e+00, 4.397000e+00, 1.694484e+06],
       [5.373800e+01, 4.392000e+00, 3.064000e+00, 1.335124e+06],
       [5.372800e+01, 4.393000e+00, 3.177000e+00, 1.281226e+06]])
```

The first list is heavy atom, the second list is hydrogen atom connected with heavy atom, the third list is H hydrogen atom connected with N, and the fourth list is peak intensity (volume). If there are multiple NOESY spectra, they need to be divided into multiple pkl files for independent storage. Currently, only 3D-NOESY spectral data is supported.

**Chemical shift table** : The file name is' chemical_shift_aligned.pkl ', and the data must be converted to 5 numpy.ndarray arrays and stored in.pkl file format as shown in the following example:

``` log
['HA' 'HB1' 'HB2' 'HB3' 'C' 'CA' 'CB' 'H' 'HA2' 'HA3']
['H' 'H' 'H' 'H' 'C' 'C' 'C' 'H' 'H' 'H']
[  4.09   1.49   1.49   1.49 174.1   52.3   19.2    8.62   3.95   3.95]
[2 2 2 2 2 2 2 3 3 3]
['ALA' 'ALA' 'ALA' 'ALA' 'ALA' 'ALA' 'ALA' 'GLY' 'GLY' 'GLY']
```

Atomic name, atomic type, chemical shift, atomic residue number, atomic residue type, atomic residue number must be aligned with the sequence in input_path.

```log
>>>>>>>>>>>>>>>>>>>>>>repeat_idx 0, contact_info_input, 42.0, confidence 84.72035603116198, contact_pred_rate_input 0.0, prot_name 5W9F,

input_file_path:  ./megaassign/iter_1/structure/5W9F_0.pdb
 2023-04-28 10:09:59.461741
final_violations:  0.0
output_file_path:  ./megaassign/iter_1/structure_relaxed/5W9F_0.pdb

[[14, 29], [22, 23], [11, 15], [20, 31], [37, 38]]
....
```

Here is a part of log, FAAST will run multiple iterations, each iteration will run RASP model 20 times to get 20 different PDBS, repeat_idx refers to the number of PDBS in the iteration, contact_info_input is the position of effective input contact information. confidence is the pdb confidence of the round iteration, contact_pred_rate_input is the contact information that takes effect after inference, prot name is the name of the protein, and input_file_path is the input for running the relax protein.

## Runtime

image.png

The figure above compares the resolution time and accuracy of the FAAST model to the traditional model. The Ascend 910 aarch64 system, for example, is deployed in about half an hour in an environment where the hardware driver package is already installed. The single sequence FAAST model runs on average, which means that even zero-based whitewashed nmr data can be analyzed in as little as one day, which would otherwise take several months or more.

## Reference

[1] Jumper J, Evans R, Pritzel A, et al. Applying and improving AlphaFold at CASP14[J].  Proteins: Structure, Function, and Bioinformatics, 2021.

[2] Liu S, Zhang J, Chu H, et al. PSP: million-level protein sequence dataset for protein structure prediction[J]. arXiv preprint arXiv:2206.12240, 2022.

[3] Terwilliger T C, Poon B K, Afonine P V, et al. Improved AlphaFold modeling with implicit experimental information[J]. Nat Methods, 2022, 19:1376–1382.

## Acknowledgement

- [ARIA](<http://aria.pasteur.fr/documentation>)
- [ColabFold](https://github.com/sokrypton/ColabFold)
- [AlphaFold2](https://github.com/deepmind/alphafold)
- [Biopython](https://biopython.org)
- [HH Suite](https://github.com/soedinglab/hh-suite)
- [Kalign](https://msa.sbc.su.se/cgi-bin/msa.cgi)
- [ML Collections](https://github.com/google/ml_collections)
- [NumPy](https://numpy.org)
- [OpenMM](<https://github.com/openmm/openmm>)
